src/renderers/dom/server/__tests__/ReactServerRendering-test.js

```diff
@@ -254,6 +306,463 @@ describe('ReactServerRendering', function() {
+            return (<TagName type="text"
+              {...{[valueKey]: this.state[valueKey]}}
+              onChange={this.handleChange.bind(this)}
+              {...extraProps}>{children}</TagName>);
+          }
+        };
+      };
+
+      const testControlledField = (initialValue, changedValue, TagName = 'input',
+        valueKey = 'value', extraProps = {}, children = null) => {
+
+        let changeCount = 0;
+        const ControlledField = getControlledFieldClass(
+          initialValue, () => changeCount++, TagName, valueKey, extraProps, children
+        );
+
+        let field = connectToServerRendering(<ControlledField/>).firstChild;
+
+        expect(changeCount).toBe(0);
+        expect(field[valueKey]).toBe(initialValue);
+
+        // simulate a user typing.
+        field[valueKey] = changedValue;
+        ReactTestUtils.Simulate.change(field);
+
+        expect(changeCount).toBe(1);
+        expect(field[valueKey]).toBe(changedValue);
+      };
+
+      it('should render a controlled text input',
+        () => testControlledField('Hello', 'Goodbye'));
+
+      it('should render a controlled textarea',
+        () => testControlledField('Hello', 'Goodbye', 'textarea'));
+
+      it('should render a controlled checkbox',
+        () => testControlledField(true, false, 'input', 'checked', {type:'checkbox'}));
+
+      it('should render a controlled select',
+      () => testControlledField('B', 'A', 'select', 'value', {},
+        [
+          <option key="1" value="A">Option A</option>,
+          <option key="2" value="B">Option B</option>,
+        ]));
+
+      // User interaction before client markup reconnect
+      const testFieldWithUserInteractionBeforeClientRender = (
+        element, initialValue = 'foo', changedValue = 'bar', valueKey = 'value'
+      ) => {
+        const root = renderOnServer(element);
+        const field = root.firstChild;
+        expect(field[valueKey]).toBe(initialValue);
+
+        // simulate a user typing in the field **before** client-side reconnect happens.
+        field[valueKey] = changedValue;
+
+        // reconnect to the server markup.
+        renderOnClient(element, root);
+
+        // verify that the input field was not replaced.
+        expect(root.firstChild).toBe(field);
+        expect(field[valueKey]).toBe(changedValue);
+      };
+
+      it('should not blow away user-entered text on successful reconnect to an uncontrolled input', () => {
+        testFieldWithUserInteractionBeforeClientRender(<input defaultValue="foo"/>, 'foo', 'bar');
+      });
+
+      it('should not blow away user-entered text on successful reconnect to an controlled input', () => {
+        let changeCount = 0;
+        const Component = getControlledFieldClass('foo', () => changeCount++);
+        testFieldWithUserInteractionBeforeClientRender(<Component/>, 'foo', 'bar');
+        // TODO: is this right? should onChange fire when a user modifies before markup reconnection?
```

Comment:
I genuinely don't know what the correct behavior in this case should be. Given that the user has modified the value since the input was loaded, I think we probably should queue up a change event. However, the current code does not do this, since it is currently not able to detect when the user has changed the value before client-side render. Thoughts?